### PR TITLE
vmap fallback kernel

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42324 vmap: temporarily disable support for (most) random functions
* #42248 Implement batching rules for some view ops
* **#41943 vmap fallback kernel**

If an operator doesn't have a batching rule implemented then we fallback
to this implementation. The fallback only works on out-of-place operators
that return only tensors with new memory. (e.g., no in-place operators,
no view operations).

The fallback effectively takes all of the BatchedTensors in `stack`,
slices them, and runs `op` on all of the corresponding slices to produce slices
of the outputs. The output slices then get `torch.stack`ed to create the
final returns.

The performance of the fallback is not very good because it introduces
an extra copy from stacking the sliced outputs. Because of this, we prefer
to write batching rules for operators whenever possible.

In the future, I'd like to disable the fallback kernel for random
functions until we have a better random story for vmap. I will probably
add a blocklist of operators to support that.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22764103](https://our.internmc.facebook.com/intern/diff/D22764103)

### CODE REVIEW

File: aten/src/ATen/BatchedFallback.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
```
Comment (zou3519) [inline aten/src/ATen/BatchedFallback.cpp:44]: Is this morally OK? `torch::jit::Stack` is typedef'ed to be `vector<ivalue>`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/stack.h#L11 and I am using it as a vector throughout this function.

There are a lot of convenience functions for `torch::jit::Stack` (https://github.com/pytorch/pytorch/blob/b85df3709a9e92f88d2e86b4860571010bc06e62/aten/src/ATen/core/stack.h#L29-L49) but they're harder to use in this use case than the `std::vector` API.

File: aten/src/ATen/BatchedFallback.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    if (idx != *batched_tensor_inputs_pos_iter) {
+      // ivalue isn't a BatchedTensor
+      pushToEachStack(ivalue);
+      continue;
+    }
+    // ivalue is a BatchedTensor
+    const auto& physical_view_for_ivalue = *input_physical_views_iter;
+    for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+      auto index = computeIndex(linear_idx, batch_sizes);
```
Comment (zou3519) [inline aten/src/ATen/BatchedFallback.cpp:95]: Instead of computing the index every time, I could modify the index in-place and add 1 to it. Is this worth it? In terms of performance it looks like it is more efficient

File: aten/src/ATen/BatchedFallback.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  // Stack the tensors together to form the result.
+  stack->clear();
+  std::vector<Tensor> output_shards;
+  for (const auto& stack : unbatched_stacks) {
+    TORCH_INTERNAL_ASSERT(stack.size() == 1)
+    output_shards.push_back(stack[0].toTensor());
+  }
+  auto flat_output = at::stack(output_shards);
```
Comment (zou3519) [inline aten/src/ATen/BatchedFallback.cpp:116]: The `stack` means that the fallback kernel won't be as performant as a user deciding to use a for-loop rather than vmap. However, the long-term state for vmap is to not rely on the fallback (and instead have batching rules defined on the operations) so I am not too worried about the performance.

## Problem

Consider the following model:
```
def model(example):
    example = F.conv(example, ...)
    example = F.relu(example, ...)
    example = F.linear(example, ...)
    return tensor
```
If the user wanted to push multiple examples via a Python for-loop, they’d do the following:
```
examples = torch.randn(N, ...)
results = []
for example in examples.unbind(0):
    results.append(model(example))
output = torch.stack(results)
```
Now, let’s assume that no batching rules were implemented for F.conv, F.relu, and F.linear. The vmap “for-loop” fallback would do the following, for each operation:

* unbind the examples along the batch dimension
* run the operator for each example
* stack the results

In other words, semantically, vmap(model)(examples) is equivalent to:
```
examples = torch.stack([F.conv(example, ...) for example in examples])
examples = torch.stack([F.relu(example, ...) for example in examples])
examples = torch.stack([F.linear(example, ...) for example in examples])
```
All of the torch.stack make the vmap “for-loop” fallback slower than the python for-loop, which is not very good for vmap.

File: aten/src/ATen/BatchedFallback.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Populate `num_batches` number of torch::jit::Stack, one for each computation.
```
Comment (ezyang) [inline aten/src/ATen/BatchedFallback.cpp:91]: A more idiomatic way to do this is to have a single stack, and then push/pop elements off of it each time you do the loop. I don't know how much simpler precomputing is, but if you maintain only a single stack the amount of memory you use and the number of dynamic allocations you do is reduced. (By the way, this is why it's a stack in the first place: so we can reuse a contiguous space of memory to do calls.

File: aten/src/ATen/BatchedFallback.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:131]: The obvious fix is to not stack: instead, use `out` to write samples directly into a preallocated output. But to preallocate the output, you need a way of computing what the size of the computation will be on one output, which is not so easy to do right now. One hacky way to do it would be to run on one example, then resize the output so it's large enough to fit the rest of the examples, and then run the loop. It'll be faster than stack. Then if we have working meta functions, you can use those to compute what the output size should be after the fact.

File: aten/src/ATen/BatchedFallback.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
```
Comment (ezyang) [inline aten/src/ATen/BatchedFallback.cpp:82]: Hmm, it's a little bit of a smell that we arbitrarily read out the first view for the number of batch dims, when in fact all of the views are guaranteed to have the same value (right?) Not sure if this is actionable though.

File: aten/src/ATen/BatchedFallback.cpp (hunk 4 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:91]: I'm not sure if this is more complicated to do or not, but I don't feel it should be that much more complicated. Now you are iterating over batch, and then for each batch, you construct the correct arguments for running the batch and then make the call. In fact, the current version of the code feels "transposed" to me.

File: aten/src/ATen/BatchedFallback.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:110]: I don't think I care too much, either way.

File: test/test_vmap.py [hunk 6]
```diff
@@ -433,5 +445,62 @@ def foo(x):
...
+        # big batch size (total 10000)
+        x = torch.randn(100, 10, 10, 5)
+        y = torch.randn(100, 10, 10)
+        result = vmap(vmap(vmap(torch.sub)))(x, y)
+        self.assertEqual(result, x - y.view(100, 10, 10, 1))
+
+    def test_fallback_masked_fill(self):
+        # NB: One day we will implement a batching rule for index_add
+        # If/when we do, this test should be replaced to test the fallback
+        # path on another operator to avoid bitrot.
```
Comment (ezyang) [inline test/test_vmap.py:484]: It would be better if you could detect if this situation occurred :) (Maybe you could use one of the apis used in `test_dispatch.py`, or a new API, to test that `index_add` doesn't have an explicit VMap key installed. Not sure it's worth the faffing about.)

Comment (zou3519) [reply test/test_vmap.py:484]: We can detect it actually -- the fallback emits a warning message and we test for the warning message in `test_fallback_masked_fill`. If someone adds a batching rule for `masked_fill`, then the test will fail and the developer will read this comment :)

File: aten/src/ATen/BatchedFallback.cpp (hunk 5 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:82]: Yes, all of the views have the same `numBatchDims()`. I'll think about if we can fix this somehow but my initial thought is that we might have to change the VmapTransforms API

File: aten/src/ATen/BatchedFallback.cpp (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:131]: We can't actually use the `out=` variant of the function because they doesn't support autograd. Because vmap occurs above autograd in the current design, vmap must desugar into operations that are compatible with autograd.

File: aten/src/ATen/BatchedFallback.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:91]: Hmm, let me think about that

File: aten/src/ATen/BatchedFallback.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:131]: RIP. cc @albanD

File: aten/src/ATen/BatchedFallback.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:110]: I'll stick with the current approach (re-compute the index vector) for now and optimize later if necessary

File: aten/src/ATen/BatchedFallback.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:59]: Answered: morally incorrect.

File: aten/src/ATen/BatchedFallback.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:91]: Yes, you're right -- the logic in this PR is "transposed". Transposing it again wasn't very difficult

File: aten/src/ATen/BatchedFallback.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+  TORCH_WARN("Batching rule not implemented for ", op.schema(), " falling back "
+             "to slow (for loop and stack) implementation");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
```
[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp]: fixed using a more idiomatic `torch::jit::last(stack, n)`

File: aten/src/ATen/BatchedFallback.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Strategy: For each batch, we are going to push slices (where applicable)
+  // of the arguments onto `working_stack`, call `op`, and store the result in
+  // `output_shards`.
+  torch::jit::Stack working_stack;
```
Comment (zou3519) [inline aten/src/ATen/BatchedFallback.cpp:94]: @ezyang I wanted to take this a step further and reuse the input stack. However, replacing this line with `auto& working_stack = *stack` causes some difficult-to-debug error messages like "RuntimeError: isTensor() INTERNAL ASSERT FAILED at "../aten/src/ATen/core/ivalue_inl.h":117, please report a bug to PyTorch. Expected Tensor but got None", "RuntimeError: IValue is not a Scalar". Perhaps there is existing code somewhere that assumes the stack must not contain additional arguments unrelated to the current operation?

**Comment by zou3519** [comment]
I updated the strategy to do stack push-pop

File: aten/src/ATen/BatchedFallback.cpp (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/BatchedFallback.cpp:125]: Note that this will depend on the overall model: a short model (few ops) won't see much difference while a deeper model will be more impacted.
Also I guess that the output size vs complexity of the function itself will have a big impact here.

But in general, for functions with inplace versions (`foo_()`), we could do this and keep autograd. But we will need a list of such functions and register a different fallback kernel for these.
For other functions, I can't think of a better way to do it :/

File: aten/src/ATen/BatchedFallback.cpp (hunk 8 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:94]: Can you get a full backtrace (and state of the stack) from gdb here? This sounds unexpected.

File: aten/src/ATen/BatchedFallback.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:125]: @albanD what is being asked for here is a little different. Imagine:

```
x1 = torch.randn(2, requires_grad=True)
x2 = torch.randn(2, requires_grad=True)
x3 = torch.randn(2, requires_grad=True)
y = torch.empty(3, 2)
x1.add(w, out=y[0])
x2.add(w, out=y[1])
x3.add(w, out=y[2])
y.backward()
```

It would be cool if this worked.

Comment (albanD) [reply aten/src/ATen/BatchedFallback.cpp:125]: Yes it would :D 
As mentioned in the `out=` autograd proposal [here](https://github.com/pytorch/pytorch/issues/29378#issuecomment-571805569), that would work just fine.
Unfortunately, detecting these valid cases and implementing this overall is not going to be done any time soon I think.

The version for the `foo_()` variants would work with the current code base though (even though it won't work with as many ops).

File: aten/src/ATen/BatchedFallback.cpp (hunk 8 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp]: I figured out the problem, it was a bug in my code. I've changed the code to just use the input stack.

If you're interested in the bug: somewhere above the code uses `torch::jit::last` to get an `ArrayRef<IValue>` of the inputs on the stack. But it turns out that when you modify the stack by pushing elements to it, the original `ArrayRef<IValue>` becomes garbage because the underlying `std::vector` that backs the stack is allowed to reallocate memory (and it does). My code was reading garbage values and that is how the errors happened.

File: aten/src/ATen/BatchedFallback.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      // argument is a BatchedTensor
+      const auto& physical_view_for_argument = *input_physical_views_iter;
+      torch::jit::push(stack, physical_view_for_argument.tensor().index(index));
+      batched_tensor_inputs_pos_iter++;
+      input_physical_views_iter++;
+    }
+
+    op.callBoxed(stack);
+
+    // We assume there is a single tensor return
```
Comment (ezyang) [inline aten/src/ATen/BatchedFallback.cpp:137]: Hmmmmmmmm

Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:137]: This is probably not true in general

File: aten/src/ATen/BatchedFallback.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+    auto input_physical_views_iter = input_physical_views.begin();
+    for (int64_t arg_idx = 0; arg_idx < num_arguments; ++arg_idx) {
+      // The accounting for this is tricky. Assume `op` has 4 args, A, B, C, D.
+      // and `arg_idx = 2`. Every iteration of this loop pushes one element
+      // onto the stack, so we've already pushed two things (call them A' and B').
+      // At this point, the stack should look like the following:
+      // [..., A, B, C, D, A', B'] (top of stack here)
+      // To get at C (argument number 2), we have to peek `num_arguments`
+      // into the stack from the top
```
Comment (ezyang) [inline aten/src/ATen/BatchedFallback.cpp:121]: TBH, you should just remember where the top of the stack was when you entered and then just index off of that

Comment (ezyang) [reply aten/src/ATen/BatchedFallback.cpp:121]: (You don't want a from top of stack accessor here)

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/BatchedFallback.cpp (hunk 9 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:137]: Yeah, that'll get fixed in the future :) (probably one of the next three diffs)

File: aten/src/ATen/BatchedFallback.cpp (hunk 10 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:121]: Are you suggesting to store a pointer to the top of the stack? The problem with a pointer is that the memory location can change due to reallocations as you add new elements onto the stack due to how `std::vector` works

Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:121]: (answered offline): Ed's suggestion was to store an index

File: aten/src/ATen/BatchedFallback.cpp (hunk 9 above)
Comment (zou3519) [reply aten/src/ATen/BatchedFallback.cpp:133]: I tried implementing support for multiple returns but I ran into https://github.com/pytorch/pytorch/issues/42378 (there's a minimal repro in the example). Should an operator with multiple returns push all of them onto the stack?
